### PR TITLE
License sale contracts

### PR DESCRIPTION
**Pull request by Archethect**

### CODE REVIEW

**Review by eludius18 — DISMISSED** [review]
LGTM ! :)

**Review by sergey-sobko — DISMISSED** [review]
👌

**Review by eludius18 — DISMISSED** [review]

File: contracts/PlayFiLicenseSale.sol [hunk 1]
```diff
@@ -106,6 +107,10 @@ IPlayFiLicenseSale
         standardDiscountPercentage = 5;
     }
 
+    /// @notice Claims licenses for team members and make sure they do not exceed their personal claim cap.
+    /// @param amount The amount of licenses to claim
+    /// @param data Index and claimCap in encoded format
+    /// @param merkleProof The proof used to verify weather the caller is allowed to claim the licenses
```
Comment (therealdannzor) [inline contracts/PlayFiLicenseSale.sol]:
```suggestion
    /// @param merkleProof The proof used to verify whether the caller is allowed to claim the licenses
```

Comment (therealdannzor) [reply contracts/PlayFiLicenseSale.sol]: minor sp

File: contracts/PlayFiLicenseSale.sol [hunk 2]
```diff
@@ -118,6 +123,11 @@ IPlayFiLicenseSale
         emit TeamLicensesClaimed(msg.sender, amount);
     }
 
+    /// @notice Claims licenses for friends and family + make sure they do not exceed their personal claim cap and that
+    /// they paid enough.
+    /// @param amount The amount of licenses to claim
+    /// @param data Index and claimCap in encoded format
+    /// @param merkleProof The proof used to verify weather the caller is allowed to claim the licenses
```
Comment (therealdannzor) [inline contracts/PlayFiLicenseSale.sol:132]:
```suggestion
    /// @param merkleProof The proof used to verify whether the caller is allowed to claim the licenses
```

Comment (therealdannzor) [reply contracts/PlayFiLicenseSale.sol:132]: minor sp

File: contracts/PlayFiLicenseSale.sol [hunk 3]
```diff
@@ -132,6 +142,11 @@ IPlayFiLicenseSale
         emit FriendsFamilyLicensesClaimed(msg.sender, amount);
     }
 
+    /// @notice Claims licenses for early access addresses + make sure they do not exceed their personal claim cap and
+    /// that they paid enough.
+    /// @param amount The amount of licenses to claim
+    /// @param data Index and claimCap in encoded format
+    /// @param merkleProof The proof used to verify weather the caller is allowed to claim the licenses
```
Comment (therealdannzor) [inline contracts/PlayFiLicenseSale.sol]:
```suggestion
    /// @param merkleProof The proof used to verify whether the caller is allowed to claim the licenses
```

Comment (therealdannzor) [reply contracts/PlayFiLicenseSale.sol]: minor sp

File: contracts/PlayFiLicenseSale.sol [hunk 4]
```diff
@@ -146,6 +161,11 @@ IPlayFiLicenseSale
         emit EarlyAccessLicensesClaimed(msg.sender, amount);
     }
 
+    /// @notice Claims licenses for partners + make sure they do not exceed their personal claim cap and that
+    /// they paid enough.
+    /// @param amount The amount of licenses to claim
+    /// @param data Index and claimCap in encoded format
+    /// @param merkleProof The proof used to verify weather the caller is allowed to claim the licenses
```
Comment (therealdannzor) [inline contracts/PlayFiLicenseSale.sol:170]:
```suggestion
    /// @param merkleProof The proof used to verify whether the caller is allowed to claim the licenses
```

Comment (therealdannzor) [reply contracts/PlayFiLicenseSale.sol:170]: minor sp

File: contracts/PlayFiLicenseSale.sol [hunk 5]
```diff
@@ -180,26 +205,48 @@ IPlayFiLicenseSale
...
+    /// @param discount The percentage of the total price to be used as a discount
     function setReferral(string memory code, address receiver, uint256 commission, uint256 discount) public onlyReferralManager {
         _setReferral(code, receiver, commission, discount);
     }
 
+    /// @notice Sets tier details
+    /// @param ids the ids of the tiers to set
+    /// @param prices the prices of each tier to set
+    /// @param individualCaps the maximum amount of licenses that can be claimed per address for the tiers.
+    /// @param totalCaps the maximum amount of licenses that can be claimed in total for the tiers.
```
Comment (therealdannzor) [inline contracts/PlayFiLicenseSale.sol]: super nit (feel free to ignore): the explanation following the parameter is previously uppercased. Previous comments looked like so: param ids **T**he ids of the tiers to set

File: contracts/PlayFiLicenseSale.sol [hunk 6]
```diff
@@ -264,9 +330,11 @@ IPlayFiLicenseSale
...
     function _setReferral(string memory code, address receiver, uint256 commission, uint256 discount) internal {
-        referrals[code].discountPercentage == discount;
-        referrals[code].commissionPercentage == commission;
-        referrals[code].receiver == receiver;
+        if(discount > 50) revert InvalidDiscount();
+        if(commission > 50) revert InvalidCommission();
+        referrals[code].discountPercentage = discount;
+        referrals[code].commissionPercentage = commission;
+        referrals[code].receiver = receiver;
         emit referralUpdated(code, receiver, commission, discount);
```
Comment (therealdannzor) [inline contracts/PlayFiLicenseSale.sol]: q: if we expect this function to be called a lot, would it make sense to make `receiver` an indexed event in order to search for it in the logs? if there's already a way to track it offchain, or if deemed not needed, feel free to just resolve

**Review by therealdannzor — COMMENTED** [review]
LGTM and has my approval ✅ -- just minor comments and a question, all non-blocking, can be used as reference for future PRs etc.

File: contracts/PlayFiLicenseSale.sol (hunk 6 above)
Comment (Archethect) [reply contracts/PlayFiLicenseSale.sol]: Agree with this. Will change!

Comment (Archethect) [reply contracts/PlayFiLicenseSale.sol]: Seems like it was already done like this in the Interface:
`event referralUpdated(string code, address indexed receiver, uint256 commission, uint256 discount);`

**Review by therealdannzor — DISMISSED** [review]

**Review by therealdannzor — APPROVED** [review]
LGTM

**Review by eludius18 — APPROVED** [review]

**Pull request merged**